File: src/components/CellView.jl
# CellView.jl — per-cell @island (WASM-compiled)
#
# Owns ALL reactive chrome state for one cell — eye toggle (fold/unfold),
# state classes (queued/running/etc.), .stale class, runtime badge text.
# CodeMirror is passed as the `children` slot and lives as a child DOM
# node that CM manages itself.
#
# THERAPY COMPILER LIMITATIONS (audit-confirmed):
#
# 1. js() args MUST be DIRECT signal-getter results — anything computed
#    by a Julia helper resolves to literal `undefined`. So every effect
#    is `v = sig(); js("…use \$1…", v)` and all branching lives INSIDE
#    the JS string.
#
# 2. Prop→signal init only fires when ALL kwargs are Integer-typed
#    (Compile.jl line 898: `all_props_are_int`). One String kwarg →
#    every signal silently stays at 0 on hydration. So this island
#    keeps every kwarg as Int. cell_id is discovered from the DOM at
#    click time inside the inline onclick handlers; run/menu actions
#    call global IDE helpers (window._sessionsRunCell etc.).
#
# Per-cell signals are PRIVATE (not shared) — the WS bridge writes
# them by selecting `therapy-island[data-cell-id=...]` and poking
# `signal_N.value` directly (the documented HMR pattern).
#
# Loaded by Therapy.load_app! from src/components/. render_cell() looks
# this up at SSR via Main.TherapyApp.CellView.

using Therapy

# Cell state codes — must match the JS WS bridge encoding.
const CV_STATE_IDLE     = 0
const CV_STATE_QUEUED   = 1
const CV_STATE_RUNNING  = 2
const CV_STATE_DONE     = 3
const CV_STATE_ERRORED  = 4
const CV_STATE_SKIPPED  = 5

# Inline onclick: discover cell_id from the DOM at click time so we
# don't need a String kwarg (which would break Therapy's prop-init).
const _CV_RUN_ONCLICK  = "window._sessionsRunCell(this.closest('.cell-wrap').dataset.cellId)"
const _CV_MENU_ONCLICK = "window._sessionsShowCellMenu(this,this.closest('.cell-wrap').dataset.cellId)"

@island function CellView(children...;
                          initial_state::Int=CV_STATE_IDLE,
                          initial_stale::Int=0,
                          initial_runtime_ns::Int=0,
                          initial_open::Int=1)
    # ── Signal layout (KEEP THIS ORDER — JS WS bridge indexes by it) ──
    state, set_state         = create_signal(initial_state)
    is_stale, set_stale      = create_signal(initial_stale)
    runtime_ns, set_runtime  = create_signal(initial_runtime_ns)
    is_open, set_open        = create_signal(initial_open)

    # ── Effect: (state, is_stale) → .code-cell className ──
    # Merged from two effects so we don't have one writing className
    # then the other reading classList to preserve stale across rebuilds.
    # Subscribes to both signals; rebuilds className from scratch on
    # either change.
    create_effect(() -> begin
        s = state()
        st = is_stale()
        js("""
            var el=island.querySelector('.code-cell');
            if(!el)return;
            var base='code-cell relative overflow-hidden';
            if(\$1===1)base+=' cv-queued executing';
            else if(\$1===2)base+=' cv-running executing';
            else if(\$1===4)base+=' cv-errored';
            else if(\$1===5)base+=' cv-skipped';
            if(\$2)base+=' stale';
            el.className=base;
        """, s, st)
    end)

    # ── Effect: runtime_ns → .rt-badge text (formatted in JS) ──
    create_effect(() -> begin
        ns = runtime_ns()
        js("""
            var b=island.querySelector('.rt-badge');
            if(!b)return;
            var n=\$1;
            if(n<=0){b.textContent='';return;}
            var ms=n/1e6;
            if(ms<1){b.textContent=(n/1e3).toFixed(1)+'\\u00b5s';return;}
            if(ms<1000){b.textContent=ms.toFixed(1)+'ms';return;}
            var s=ms/1000;
            if(s<60){b.textContent=s.toFixed(1)+'s';return;}
            var m=s/60;
            if(m<60){b.textContent=m.toFixed(1)+'min';return;}
            b.textContent=(m/60).toFixed(1)+'hr';
        """, ns)
    end)

    # ── Effect: is_open → .cell-code-wrap visibility + .code-hidden class + WS ──
    # Single writer for fold state: inline style on .cell-code-wrap AND
    # .code-hidden on .cell-wrap. The class must move together with the
    # inline style — the CSS rule `.cell-wrap.code-hidden .cell-code-wrap
    # {display:none}` otherwise overrides a cleared inline style, leaving
    # the cell stuck hidden when re-opened. Init flag gates WS so hydration
    # doesn't fire a spurious toggle_fold.
    create_effect(() -> begin
        v = is_open()
        js("""
            var c=island.querySelector('.cell-code-wrap');
            if(c)c.style.display=\$1?'':'none';
            var wrap=island.closest('.cell-wrap');
            if(wrap){
                if(\$1) wrap.classList.remove('code-hidden');
                else wrap.classList.add('code-hidden');
            }
            if(!island._foldInit){island._foldInit=true;return;}
            var cid=wrap?wrap.dataset.cellId:'';
            if(cid&&window.TherapyWS&&TherapyWS.sendMessage){
                TherapyWS.sendMessage('notebook',{action:'toggle_fold',cell_id:cid,folded:!\$1});
            }
        """, v)
    end)

    # SSR initial values so first paint matches state before effects run.
    initial_class = "code-cell relative overflow-hidden"
    initial_state == CV_STATE_QUEUED  && (initial_class *= " cv-queued executing")
    initial_state == CV_STATE_RUNNING && (initial_class *= " cv-running executing")
    initial_state == CV_STATE_ERRORED && (initial_class *= " cv-errored")
    initial_state == CV_STATE_SKIPPED && (initial_class *= " cv-skipped")
    initial_stale == 1                && (initial_class *= " stale")
    initial_wrap_style = initial_open == 1 ? "" : "display:none"

    Div(:class => "cell-island",
        Div(:class => "cell-eye",
            :on_click => () -> set_open(1 - is_open()),
            Div(:style => "position:relative;width:14px;height:14px;",
                RawHtml("""<svg width="14" height="14" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.5"><path d="M17.94 17.94A10.07 10.07 0 0112 20c-7 0-11-8-11-8a18.45 18.45 0 015.06-5.94M9.9 4.24A9.12 9.12 0 0112 4c7 0 11 8 11 8a18.5 18.5 0 01-2.16 3.19M1 1l22 22"/></svg>"""),
                Show(is_open) do
                    Div(:style => "position:absolute;inset:0;background:var(--panel-bg);",
                        RawHtml("""<svg width="14" height="14" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.5"><path d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"/><circle cx="12" cy="12" r="3"/></svg>"""))
                end)),
        Div(:class => "cell-code-wrap", :style => initial_wrap_style,
            Div(:class => initial_class,
                Div(:class => "cell-ctrls absolute top-1 right-1.5 flex items-center z-10",
                    Span(:class => "rt-badge", ""),
                    Therapy.Button(:class => "ctrl-btn run-btn",
                        :title => "Run cell (Shift+Enter)",
                        :on_click => _CV_RUN_ONCLICK,
                        RawHtml("""<svg width="10" height="10" viewBox="0 0 16 16" fill="currentColor"><path d="M4 2.5v11l10-5.5z"/></svg>""")),
                    Therapy.Button(:class => "ctrl-btn menu-btn",
                        :title => "Cell actions",
                        :on_click => _CV_MENU_ONCLICK,
                        RawHtml("""<svg width="11" height="11" viewBox="0 0 16 16" fill="none" stroke="currentColor" stroke-width="1.5"><circle cx="8" cy="3" r="1.2"/><circle cx="8" cy="8" r="1.2"/><circle cx="8" cy="13" r="1.2"/></svg>"""))),
                children...)))
end

File: static/islands/CellView.js
(function() {
  window.TherapyHydrate = window.TherapyHydrate || {};
  function hydrate_cellview() {
    document.querySelectorAll('[data-component="cellview"]:not([data-hydrated])').forEach(function(island) {
      var props = JSON.parse(island.dataset.props || '{}');
      var _wb = new Uint8Array([0,97,115,109,1,0,0,0,1,23,5,96,0,0,96,1,126,0,96,3,111,111,127,0,96,2,126,126,0,96,0,1,127,2,62,4,3,100,111,109,9,115,104,111,119,95,115,119,97,112,0,2,6,101,102,102,95,106,115,5,101,102,102,95,48,0,3,6,101,102,102,95,106,115,5,101,102,102,95,49,0,1,6,101,102,102,95,106,115,5,101,102,102,95,50,0,1,3,9,8,0,0,0,0,4,0,1,0,4,5,1,112,1,4,4,6,76,15,126,1,66,0,11,126,1,66,0,11,126,1,66,0,11,126,1,66,1,11,111,1,208,114,11,111,1,208,114,11,127,1,65,127,11,127,1,65,0,11,126,1,66,0,11,126,1,66,0,11,126,1,66,0,11,126,1,66,0,11,126,1,66,0,11,111,1,208,114,11,127,1,65,127,11,7,243,1,21,8,115,105,103,110,97,108,95,48,3,0,8,115,105,103,110,97,108,95,49,3,1,8,115,105,103,110,97,108,95,50,3,2,8,115,105,103,110,97,108,95,51,3,3,4,104,107,95,50,3,4,4,104,107,95,52,3,5,12,95,114,116,95,111,98,115,101,114,118,101,114,3,6,9,95,114,116,95,98,97,116,99,104,3,7,11,95,114,116,95,112,101,110,100,105,110,103,3,8,10,95,114,116,95,115,117,98,115,95,48,3,9,10,95,114,116,95,115,117,98,115,95,49,3,10,10,95,114,116,95,115,117,98,115,95,50,3,11,10,95,114,116,95,115,117,98,115,95,51,3,12,3,95,104,49,0,4,9,95,101,102,102,101,99,116,95,48,0,5,9,95,101,102,102,101,99,116,95,49,0,6,9,95,101,102,102,101,99,116,95,50,0,7,12,95,115,104,111,119,95,98,97,114,101,95,52,0,8,12,95,115,104,111,119,95,52,95,102,114,97,103,3,13,9,95,114,116,95,102,108,117,115,104,0,10,4,95,104,119,49,0,11,9,10,1,0,65,0,11,4,9,5,6,7,10,184,2,8,9,0,66,1,35,3,125,36,3,11,46,0,35,6,65,0,78,4,64,35,9,66,1,35,6,173,134,132,36,9,11,35,6,65,0,78,4,64,35,10,66,1,35,6,173,134,132,36,10,11,35,0,35,1,16,1,11,25,0,35,6,65,0,78,4,64,35,11,66,1,35,6,173,134,132,36,11,11,35,2,16,2,11,25,0,35,6,65,0,78,4,64,35,12,66,1,35,6,173,134,132,36,12,11,35,3,16,3,11,5,0,35,3,167,11,50,1,1,127,35,6,65,0,78,4,64,35,12,66,1,35,6,173,134,132,36,12,11,35,3,167,65,0,71,34,0,35,14,70,4,64,15,11,32,0,36,14,35,5,35,13,32,0,16,0,11,82,2,1,127,1,126,3,64,32,1,65,4,72,4,64,32,0,32,1,173,136,66,1,131,167,4,64,32,1,36,6,66,126,32,1,173,137,34,2,35,9,131,36,9,32,2,35,10,131,36,10,32,2,35,11,131,36,11,32,2,35,12,131,36,12,32,1,17,0,0,11,32,1,65,1,106,33,1,12,1,11,11,11,61,1,1,126,35,7,65,1,106,36,7,16,4,35,7,65,0,74,4,64,35,8,35,12,132,36,8,11,35,7,65,1,107,36,7,35,7,69,4,64,3,64,35,8,80,69,4,64,35,8,66,0,36,8,16,10,12,1,11,11,11,11]);
      var _io = __tw.io(island);
      var _eff={};
      _io.eff_js={eff_0:function(_p0,_p1){if(_eff[0])_eff[0](_p0,_p1);},eff_1:function(_p0){if(_eff[1])_eff[1](_p0);},eff_2:function(_p0){if(_eff[2])_eff[2](_p0);}};
      WebAssembly.instantiate(_wb, _io).then(function(result) {
        var ex = result.instance.exports;
        island._wasmExports = ex;
        island.dataset.hydrated = "true";
        if (props.initial_state !== undefined && typeof props.initial_state === 'number') ex.signal_0.value = BigInt(props.initial_state);
        if (props.initial_stale !== undefined && typeof props.initial_stale === 'number') ex.signal_1.value = BigInt(props.initial_stale);
        if (props.initial_runtime_ns !== undefined && typeof props.initial_runtime_ns === 'number') ex.signal_2.value = BigInt(props.initial_runtime_ns);
        if (props.initial_open !== undefined && typeof props.initial_open === 'number') ex.signal_3.value = BigInt(props.initial_open);
        var hk_2 = island.querySelector('[data-hk="2"]');
        ex.hk_2.value = hk_2;
        var hk_4 = island.querySelector('[data-hk="4"]');
        ex.hk_4.value = hk_4;
        _eff[0]=function(_p0,_p1){    var el=island.querySelector('.code-cell');
    if(!el)return;
    var base='code-cell relative overflow-hidden';
    if(Number(_p0)===1)base+=' cv-queued executing';
    else if(Number(_p0)===2)base+=' cv-running executing';
    else if(Number(_p0)===4)base+=' cv-errored';
    else if(Number(_p0)===5)base+=' cv-skipped';
    if(Number(_p1))base+=' stale';
    el.className=base;
;};
        _eff[1]=function(_p0){    var b=island.querySelector('.rt-badge');
    if(!b)return;
    var n=Number(_p0);
    if(n<=0){b.textContent='';return;}
    var ms=n/1e6;
    if(ms<1){b.textContent=(n/1e3).toFixed(1)+'\u00b5s';return;}
    if(ms<1000){b.textContent=ms.toFixed(1)+'ms';return;}
    var s=ms/1000;
    if(s<60){b.textContent=s.toFixed(1)+'s';return;}
    var m=s/60;
    if(m<60){b.textContent=m.toFixed(1)+'min';return;}
    b.textContent=(m/60).toFixed(1)+'hr';
;};
        _eff[2]=function(_p0){    var c=island.querySelector('.cell-code-wrap');
    if(c)c.style.display=Number(_p0)?'':'none';
    var wrap=island.closest('.cell-wrap');
    if(wrap){
        if(Number(_p0)) wrap.classList.remove('code-hidden');
        else wrap.classList.add('code-hidden');
    }
    if(!island._foldInit){island._foldInit=true;return;}
    var cid=wrap?wrap.dataset.cellId:'';
    if(cid&&window.TherapyWS&&TherapyWS.sendMessage){
        TherapyWS.sendMessage('notebook',{action:'toggle_fold',cell_id:cid,folded:!Number(_p0)});
    }
;};
        var _show_4_frag = document.createDocumentFragment();
        while(hk_4.firstChild) _show_4_frag.appendChild(hk_4.firstChild);
        hk_4.style.display = '';
        ex._show_4_frag.value = _show_4_frag;
        hk_2.$$click = function(e){ex._hw1();};
        island.addEventListener("click", function(e){var el=e.target;while(el&&el!==island){if(el.$$click){el.$$click(e);return;}el=el.parentNode;}});
        ex._rt_flush(BigInt(15));
      }).catch(function(e){console.error('[therapy] WASM instantiation failed for cellview:',e);});
    });
  }
  window.TherapyHydrate["cellview"] = hydrate_cellview;
  if (!window._therapyRouterHydrating) (window.requestIdleCallback||setTimeout)(hydrate_cellview);
})();